### PR TITLE
add job garbage collector support

### LINKED ISSUE #64470: Job objects are not cleaned up
/kind bug
/sig apps

It's easy to make jobs and forget to delete them. If you make enough of them, you can overload a control plane (which is typically sized to support the number of nodes and not an arbitrarily high number of jobs).

We have a "podgc" for deleting terminated pods if too many of them build up, so there is precedent.

### PR DESCRIPTION
**Pull request by ceshihao**
**What this PR does / why we need it**:
It's easy to make jobs and forget to delete them. I add jobgc controller similar to `podgc`.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #64470

**Special notes for your reviewer**:
cc @lavalamp 

**Release note**:

```release-note
Support finished job garbage collecting.
```

### CODE REVIEW

**Review by yue9944882 — CHANGES_REQUESTED** [review]
IMHO, job gc should be based on watches by informer instead of a constant checking period and corresponding events should also be recorded on gc.

**Comment by lavalamp** [comment]
/assign @janetkuo 

I am OK with a second one-off GC like this. I don't consider this a good ultimate solution, but it's OK for now. I don't think this concept applies to resources other than pods and jobs?

File: pkg/controller/jobgc/gc_controller.go [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+func (gcc *JobGCController) gcFinished(jobs []*v1.Job) {
+	finishedJobs := []*v1.Job{}
+	for _, job := range jobs {
+		if jobctrl.IsJobFinished(job) {
+			finishedJobs = append(finishedJobs, job)
+		}
+	}
+
+	finishedJobCount := len(finishedJobs)
+	sort.Sort(byCreationTimestamp(finishedJobs))
```
[the remark this answers is not in the document]
Comment (yue9944882) [reply pkg/controller/jobgc/gc_controller.go]: No, an anonymous func should be enough without extra type definition and function definition.

**Comment by ceshihao** [comment]
i see.

File: pkg/controller/jobgc/gc_controller.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ceshihao) [reply pkg/controller/jobgc/gc_controller.go]: I see. Thank you.

**Comment by janetkuo** [comment]
Please hold off implementing another GC controller following the podgc pattern. Podgc has its own problem, it's designed that way because it was created before generic garbage collector. Now that we have generic garbage collector I'd like to solve Job GC in a more generic way. See my proposal: https://github.com/kubernetes/community/pull/2287

**Review by janetkuo — CHANGES_REQUESTED** [review]
Please hold off implementing another GC controller following the podgc pattern. Podgc has its own problem, it's designed that way because it was created before generic garbage collector. Now that we have generic garbage collector I'd like to solve Job GC in a more generic way. See my proposal: https://github.com/kubernetes/community/pull/2287

**Pull request closed without merging**